src/Client/DiscoveryCache.cs

```diff
@@ -54,14 +56,22 @@ public DiscoveryCache(string authority, Func<HttpMessageInvoker> httpClientFunc,
     /// Get the DiscoveryResponse either from cache or from discovery endpoint.
     /// </summary>
     /// <returns></returns>
-    public Task<DiscoveryDocumentResponse> GetAsync()
+    public async Task<DiscoveryDocumentResponse> GetAsync()
     {
-        if (_nextReload <= DateTime.UtcNow)
+        await _semaphoreSlim.WaitAsync()
```

Comment:
We'd like to have a timeout for the semaphore's wait. We'd need to figure out where that would come from though.